### ISSUE TITLE
OSX Build server is unreliable

### ISSUE THREAD

**Issue by remyroy** [CleverRaven/Cataclysm-DDA#19297]
The current directory for experimental for OSX builds is empty. New OSX users cannot download the game. @narc0tiq Would it be possible to prevent deleting the builds and keep like the last 5 or 10 instead of deleting them from the destination directory when new ones cannot be build?

Pinging @patternoia for more info and insight on this process.

**Comment by patternoia** [comment]
Yes, I had to put down the build server, because I run it on a private laptop and I needed to take it with me on a trip.

We could resolve the situation in following way:
1. There is in jenkins a mechanism for build promotion to publish the builds in these directories `http://dev.narc.ro/cataclysm/jenkins-promoted/` It might demand further jenkins configuration.
2. I could every now and then make the .dmg builds manually and publish them in their respective directories `http://dev.narc.ro/cataclysm/jenkins-latest/OSX/` For this I would need either the means to auth by public key from build server or the pass-phrase for private key I already have for respective deployment ssh-user on the jenkins server from you @narc0tiq 
3. Investigate the possibility to cross compile C:DDA from linux to mac with these tools `https://github.com/tpoechtrager/osxcross` I have succeeded to build some software with it on my Ubuntu VM but never succeeded to fully build C:DDA due to lack of time.

**Comment by narc0tiq** [comment]
The current old build deleter is just this crontab bit:

```crontab
# m h  dom mon dow   command
33 2 * * * find /var/lib/jenkins/cataclysm-builds -mtime +2 -type f -delete
```

I could change it, of course -- ideally, it should be a script that checks number of files in the directory and ignores it if it's less than 5. I'll look into that, I'm sure it's doable.

The promoted builds are primarily for stable releases. We were investigating the possibility of using git tags or some other marker to tell Jenkins that certain builds are stable (as in, keep forever), but then of course we'd be promoting something before it's built -- what if the build fails, or the binary turns out to have a surprise bug that makes it unplayable? I'd want some affirmative claim by a authorized person who can test the build *before* triggering the promo.

**Comment by remyroy** [comment]
I'm not a bash expert [but I can google](https://stackoverflow.com/questions/25785/delete-all-but-the-most-recent-x-files-in-bash). Something like this should keep only the 5 most recent files in the `/var/lib/jenkins/cataclysm-builds/jenkins-latest/OSX/Tiles/` directory (assuming that is the proper directory for those builds and that you have some kind of GNU version of `xargs`) :

```bash
ls -tp /var/lib/jenkins/cataclysm-builds/jenkins-latest/OSX/Tiles/ | grep -v '/$' | tail -n +6 | xargs -d '\n' rm --
```

You would have to add a separate cron task for each directories. If you want to keep more than 5, adjust the tail parameter.

**Comment by narc0tiq** [comment]
That looks great, I'll plug it in now and it'll be ready for the future.

**Comment by narc0tiq** [comment]
The end result, if anyone's curious (also visible at <http://dev.narc.ro/cataclysm/jenkins-latest/delete-olden-builds.sh>):
```bash
for i in ~cdda-deploy/cata-builds/*/*; do
    find $i -mindepth 1 -maxdepth 1 -printf '%Ts %p\n' | sort -n | cut -d ' ' -f2- | tail -n +6 | xargs -d '\n' rm -f --
done
```

(via <http://unix.stackexchange.com/a/49540>)

**Comment by remyroy** [comment]
There seems to be issues with jenkins and the build process. I don't know if this is related but [I can see build #5883 completed](http://ci.narc.ro/view/Cataclysm-DDA/job/Cataclysm-Matrix/changes) but it is not available in [the windows tiles directory](http://dev.narc.ro/cataclysm/jenkins-latest/Windows/Tiles/) for instance.

Is it possible the new cron task keep the oldest 5 files instead of the most recent ones?

**Comment by remyroy** [comment]
This might be related to #19314

**Comment by remyroy** [comment]
@patternoia Is your OSX build slave back? It would be nice if you could bring it back just for one last successful build that could be kept with this new cron setup. This way new Mac user could at least download something.

**Comment by patternoia** [comment]
@remyroy not until the end of the week. I discussed manual uploads with narc privately, but there are currently some issues.
In the meanwhile I have brought up my second build machine, though it is unstable. Let's see if we will have some binaries uploaded from it.

**Comment by narc0tiq** [comment]
> Is it possible the new cron task keep the oldest 5 files instead of the most recent ones?

Of course it is -- it needed `head -n -5` instead of `tail -n +6`. Fixed now. Also, @patternoia, I peeked at the auth.log and maybe I found what it was complaining about.

**Comment by patternoia** [comment]
So, as I expected the second build server is still unstable. 
I have no idea why, but it goes offline like every half an hour, which is especially bad while uploading the binaries to the server. This is always treated fatal and there is no chance to reupload the binaries, since jenkins' build workspace is getting wiped, ultimately resulting in broken partial uploads.

I did now manually put the fresh binaries on the server and will continue doing so every now and then. Many thanks to @narc0tiq

**Comment by remyroy** [comment]
It seems like we are missing the most recent build in the various directories. I do not know if this is a problem with Jenkins or the new cron script.

[Build 5886](http://ci.narc.ro/view/Cataclysm-DDA/job/Cataclysm-Matrix/changes) should be completed. However, the directories where the cron task was executed ([Windows Tiles](http://dev.narc.ro/cataclysm/jenkins-latest/Windows/Tiles/), [Windows Console](http://dev.narc.ro/cataclysm/jenkins-latest/Windows/Curses/), [Linux Tiles](http://dev.narc.ro/cataclysm/jenkins-latest/Linux/Tiles/)) do not have it. Also, it seems the cron task is not executed in [Linux Console](http://dev.narc.ro/cataclysm/jenkins-latest/Linux/Curses/) directory since it has 6 builds now including the 5886 one compared to the others ones who only have 5.

This might be a off-by-one error with the new cron script or a problem with the build process itself.

**Comment by patternoia** [comment]
@remyroy  @narc0tiq It was my fault. I've misconfigured Jenkins yesterday by putting osx build servers offline, but not excluding the osx jobs from build matrix. My apologies to everyone.
Now the job is done, and all the binaries are on place.

**Comment by narc0tiq** [comment]
> it seems the cron task is not executed in Linux Console directory since it has 6 builds now

The cron task, as previously shown, executes daily at 02:33 (UTC, because that's the server's timezone). It has not run yet (it will in about 10 hours).

**Comment by narc0tiq** [comment]
For further reference, I've set it to keep 10 builds instead of 5. It seemed appropriate.

**Comment by patternoia** [comment]
#19320 should be the ultimate answer to OSX build stability issue. Please review

**Comment by remyroy** [comment]
This should mostly be fixed with work done here and in #19559 .
